libmamba/include/mamba/download/mirror.hpp

```diff
@@ -0,0 +1,127 @@
+#include <string>
+#include <vector>
+
+#include <fmt/core.h>
+
+#include "mamba/download/request.hpp"
+
+namespace mamba::download
+{
+    class MirrorID
+    {
+    public:
+
+        explicit MirrorID(std::string_view v);
+
+        std::string to_string() const;
+
+        friend bool operator<(const MirrorID& lhs, const MirrorID& rhs);
+        friend bool operator==(const MirrorID& lhs, const MirrorID& rhs);
+
+    private:
+
+        std::string m_value;
+    };
+
+    struct MirrorRequest : RequestBase
+    {
+        using header_list = std::vector<std::string>;
+
+        std::string url;
+        std::optional<header_list> headers;
+
+        MirrorRequest(
+            const RequestBase& base,
+            std::string_view url,
+            std::optional<header_list> headers = std::nullopt
+        );
+
+        ~MirrorRequest() = default;
+        MirrorRequest(const MirrorRequest&) = default;
+        MirrorRequest& operator=(const MirrorRequest&) = default;
+        MirrorRequest(MirrorRequest&&) = default;
+        MirrorRequest& operator=(MirrorRequest&&) = default;
+    };
+
+    class Mirror
+    {
+    public:
+
+        using request_generator = std::function<MirrorRequest(const Request&, const Content*)>;
+        using request_generator_list = std::vector<request_generator>;
+
+        virtual ~Mirror() = default;
+
+        Mirror(const Mirror&) = delete;
+        Mirror& operator=(const Mirror&) = delete;
+        Mirror(Mirror&&) = delete;
+        Mirror& operator=(Mirror&&) = delete;
+
+        const MirrorID& id() const;
+        request_generator_list get_request_generators() const;
+
+        std::size_t max_retries() const;
+        std::size_t successful_transfers() const;
+        std::size_t failed_transfers() const;
+
+        bool can_accept_more_connections() const;
+        bool can_retry_with_fewer_connections() const;
+
+        void cap_allowed_connections();
+        void increase_running_transfers();
+        void update_transfers_done(bool success, bool record_success);
+
+    protected:
+
+        Mirror(MirrorID id, std::size_t max_retries = 3);
+
+    private:
+
+        virtual request_generator_list get_request_generators_impl() const = 0;
+
+        MirrorID m_id;
+        size_t m_max_retries;
+
+        // TODO: use sychronized value
+        std::mutex m_stats_mutex;
+        std::optional<std::size_t> m_allowed_connections = std::nullopt;
+        std::size_t m_max_tried_connections = 0;
+        std::size_t m_running_transfers = 0;
+        std::size_t m_successful_transfers = 0;
+        std::size_t m_failed_transfers = 0;
+    };
+
+    // This class will be renamed FileMirror when
+    // other mirrors have been plugged. It is used
+    // for everything to ensure a smooth transition
+    class PassThroughMirror : public Mirror
+    {
+    public:
+
+        using request_gerenrator_list = Mirror::request_generator_list;
+
+        PassThroughMirror();
+
+        static MirrorID id();
```

Comment:
Each mirror is responsible for creating a `MirrorID` based on the arguments it is given (most of the time, it will be passed an URL only). In the case of this Mirror, passing an argument does not really make sense, since there should be a single PassThroughMirror only. I will rename it into `make_id`.